Top Posts Widget: new filters

ryanhellyer opened the pull request:
We need to implement some extra features into the Jetpack Top Posts widget. To avoid us forking the code, we'd appreciate if some extra hooks and filters could be added to enable us to do this.

We need to add (and save) extra fields, and to be able to modify the number of days the statistics are set for on a per widget basis (the current filter limits us to only changing the number of days across all widgets in use on a page).

#### Changes proposed in this Pull Request:
Adding ability to:
1. Inject extra fields (and save them) into the top posts widget.
2. Filter the number of days on a per-widget basis.

jeherve commented:
Could you [add inline documentation](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters) to all actions and filters you've added, including `@module widgets`, `@since 4.0.0`, and the relevant `@param`? [Here is an example](https://github.com/ryanhellyer/jetpack/blob/master/modules/widgets/top-posts.php#L191).

Since you edited the `jetpack_top_posts_days` filter, you'll also need to update the inline docs to include the new parameter, as well as the new `@since`.

Thanks!

ryanhellyer commented:
Thanks @jeherve :) I hadn't added those docs as most of the existing ones were missing them.

I've added inline documentation now for the relevant filters and hooks, plus I've modified the existing inline documentation for the jetpack_top_posts_days filter.

I have not modified any of the existing functions inline documentation where they have no documentation. If you would like me to do that, I can do that for too.

jeherve commented:
> I hadn't added those docs as most of the existing ones were missing them.

All hooks (`apply_filters`, `do_action`) in Jetpack are documented. If you find a hook that's not documented, let us know!

> I've added inline documentation now for the relevant filters and hooks, plus I've modified the existing inline documentation for the jetpack_top_posts_days filter.

Thanks!

> I have not modified any of the existing functions inline documentation where they have no documentation. If you would like me to do that, I can do that for too.

No, that's alright. Although it can be useful, documenting functions is not a requirement when adding new code to Jetpack.

jeherve commented on modules/widgets/top-posts.php:
```diff
@@ -124,11 +133,18 @@ function form( $instance ) {
+		</p><?php
 
-		<p><?php esc_html_e( 'Top Posts &amp; Pages by views are calculated from 24-48 hours of stats. They take a while to change.', 'jetpack' ); ?></p>
+		do_action( 'jetpack_widget_top_posts_after_fields', array( $instance, $this ) );
```
Could you document this action hook as well?

Thanks!

ryanhellyer replied:
Sorry, I was thinking the documentation was for the add_action, not do_action hence I missed this. This has been added now.

Should I perhaps remove the documentation that I added for the add_action call?

jeherve replied:
No, that's alright, it doesn't hurt to leave it there! :)

Thank you!

jeherve commented:
Apart from that last documentation bit, this is good to merge!

defries commented:
Thanks for a smooth integration here @jeherve. Made so much more sense for us to hook into existing code than to build from scratch. Cheers!
